src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,241 @@
-
-        Long productsInCart = productList.stream().filter(p -> p.getProductName().equals(productName)).count();
 
-
-
-        if (productsInCart == 0) {
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
+        Cart cart;
+        if (cartRepository.findByUserId(user.getId()) != null) {
+            cart = cartRepository.findByUserId(user.getId());
+        } else {
+            throw new EntityNotFoundException("Cart not found in the Database!");
         }
 
-        double productTotalRemove = 0;
+        List<CartEntry> cartEntries;
+        Integer initialQuantity = 0;
 
-        List<Product> products = productList.stream().filter(p -> p.getProductName().equals(productName)).collect(Collectors.toList());
-        productList.removeAll(products);
+        cartEntries = cart.getEntries();
+        CartEntry cartEntry = cartEntries.stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().orElse(null);
 
-        cart.setTotal(cart.getTotal() - (products.size() * product.getProductPrice()));
-
-        for (int i = 0; i < quantity; i++) {
-            productList.add(product);
-            productTotalRemove = product.getProductPrice() + productTotalRemove;
+        if (cartEntry == null) {
+            throw new EntityNotFoundException("CartEntry not found in the Database!");
+        } else {
+            initialQuantity = cartEntry.getQuantity();
+            cartEntry.setQuantity(quantity);
+            cartEntry.setCart(cart);
         }
-        cart.setTotal(cart.getTotal() + productTotalRemove);
 
-        cart.setUser(user);
+        cart.setEntries(cartEntries);
 
-        cart.setProductList(productList);
+        double total = cart.getTotal() - (initialQuantity * productDTO.getPrice()) + (quantity * productDTO.getPrice());
 
-        cartRepository.save(cart);
-
-        return CartMapper.cartEntityToDto(cart);
+        cart.setTotal(total);
 
+        cartRepository.save(cart);
 
+        return cartEntityToDto(cartRepository.findByUserId(user.getId()));
     }
 
-    public CartDTO removeProductFromCart(String email, String productName, int quantity)
+    public CartDTO removeProductFromCart(String email, String productName)
             throws InvalidInputDataException, EntityNotFoundException {
+        log.info("removeProductFromCart - process started");
 
-        Product product;
+        ProductDTO productDTO;
         try {
-            product = ProductMapper.DtoToProductEntity(productServiceClient.callGetProductByProductName(productName));
+            productDTO = productServiceClient.callGetProductByProductName(productName);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Product: " + productName + " not found in the Database!");
         }
 
-        User user;
+        if (productDTO.getStock() == 0) {
```

Comment:
I think you don't need this validation since you return some products to the stock. so you won't need the get product retrieval a few lines above

Reply:
Done!